### PR TITLE
[bazel] Testcases for `otbn_autogen_sim_test`

### PR DESCRIPTION
**Pull request by andrea-caforio**
The testcase feature allows OTBN tests to be instrumented with a
single HJSON file that comprises both input and output values for
registers and DMEM alike. It is an iteration on the cumbersome legacy
system that grouped input values in .s file and output values (only
for registers) in a .exp file.

With the legacy system, it was further possible to automatically
generate multiple tests through a Python script (and the
`otbn_autogen_sim_test` rule) that could optionally incorporate
randomness into the generation of its inputs. This commit extends
aforementioned Bazel rule to also generate tests that make use of the
testcase feature.

### CODE REVIEW

File: hw/ip/otbn/util/shared/testgen.py [hunk 1]
```diff
@@ -32,14 +54,43 @@ def write_test_data(inputs: Dict[str, bytes], data_file: TextIO) -> None:
...
+            "dmem": {
+            }
+        },
+        "output": {
+            "regs": {
+            },
+            "dmem": {
+            }
+        }
+    }
```
Comment (sasdf) [inline hw/ip/otbn/util/shared/testgen.py]: Just an idea, feel free to ignore if not applicable: One alternative is to compose the testcase dictionary directly within the tests. It would likely result in testgen code similar to a "templated HJSON", improving consistency with our existing handwritten HJSON test cases.

When building the test case dictionary, any unused fields (such as entrypoint or input.regs) can be omitted. The test case reader will automatically apply default empty values for any missing keys.

File: rules/otbn.bzl [hunk 2]
```diff
@@ -481,6 +494,8 @@ otbn_autogen_sim_test = rv_rule(
             cfg = "exec",
         ),
         "seed": attr.int(mandatory = True),
+        # 0: Legacy test, 1: Test with testcase
+        "type": attr.int(default = 0),
```
Comment (sasdf) [inline rules/otbn.bzl]: Can we use string choices instead (e.g. "expectation", "testcase")?

File: sw/otbn/crypto/mldsa87/tests/BUILD [hunk 3]
```diff
@@ -33,3 +34,32 @@ mldsa87_unit_tests = [
...
+        srcs = [
+            "mldsa87_ntt_test.s",
+            "//sw/otbn/crypto/mldsa87:mldsa87_ntt.s",
+        ],
+        seed = i,
+        testgen = ":mldsa87_ntt_testgen",
+        type = 1,
+    )
+    for i in range(10)
+]
```
Comment (sasdf) [inline sw/otbn/crypto/mldsa87/tests/BUILD]: Instead of using list comprehension, a macro similar to `otbn_sim_test_suite` could be used to make it prettier.

File: sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                        type=int,
+                        required=False,
+                        help=('Seed value for pseudorandomness.'))
+    parser.add_argument('hjson',
+                        metavar='FILE',
+                        type=argparse.FileType('w'),
+                        help=('Testcase HJSON file.'))
+    args = parser.parse_args()
+
+    gen_ntt_test(args.seed, args.hjson)
```
Comment (sasdf) [inline sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py]: Since the program arguments are predefined in .bzl, maybe we can move this boilerplate to shared.testgen and seed the RNG there?

File: hw/ip/otbn/util/shared/testgen.py (hunk 1 above)
Comment (andrea-caforio) [reply hw/ip/otbn/util/shared/testgen.py]: Let me see what I can do. :-)

Comment (andrea-caforio) [reply hw/ip/otbn/util/shared/testgen.py]: I tinkered a bit around. The autogen testcases are now similar to the handwritten ones, i.e., 
a template HJSON file is rendered by the Python script.

See example in the second commit.

File: sw/otbn/crypto/mldsa87/tests/BUILD (hunk 3 above)
Comment (andrea-caforio) [reply sw/otbn/crypto/mldsa87/tests/BUILD]: Done.

File: sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py (hunk 4 above)
Comment (andrea-caforio) [reply sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py]: I'm sure how this would work because this is the entry point of the program that simply
calls some functions in `shared.testgen`. How would `shared.testgen` have access to the
seed?

Comment (sasdf) [reply sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py]: For example, we could create a `shard.testgen.testcase(gen: Fn(seed) -> Dict)` helper that accepts a callback for creating testcase dict and manages the underlying setup and I/O operations.

It can then be invoked directly (or applied using the `@` decorator syntax).

File: hw/ip/otbn/util/shared/testgen.py (hunk 1 above)
Comment (sasdf) [reply hw/ip/otbn/util/shared/testgen.py]: Ahh, I'm referring to inline the dict directly in the test python code, not a real template engine. Sorry for confusion.

e.g.
```python
rendered_testcase = {
  "entrypoint": "main",
  "input": { "dmem": { "_ntt_x": _ntt_x } },
  ...
}
```

File: rules/otbn.bzl [hunk 5]
```diff
@@ -217,6 +217,21 @@ def otbn_sim_test_suite(name, tests, **kwargs):
         tags = tags,
     )
 
+def otbn_sim_testgen(name, srcs, testgen, number, testcase = None):
+    """Bazel macro to automatically generate `number` of testcases.
+
+    An index i for i <= 0 < `number` will be appended to the test name.
```
Comment (etterli) [inline rules/otbn.bzl]:
```suggestion
    An index i (0 <= i < `number`) will be appended to the test name.
```
?

File: rules/otbn.bzl [hunk 6]
```diff
@@ -235,16 +250,28 @@ def _otbn_autogen_sim_test_impl(ctx):
     value for randomness with '-s'.
```
Comment (etterli) [inline rules/otbn.bzl:260]: Should this be updated as well?

File: hw/ip/otbn/util/shared/testgen.py (hunk 1 above)
Comment (andrea-caforio) [reply hw/ip/otbn/util/shared/testgen.py]: Haha...I fixed it. :-)

File: sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py (hunk 4 above)
Comment (andrea-caforio) [reply sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py]: I didn't know about the decorator syntax. It's pretty neat. I removed the boilerplate with it.

File: rules/otbn.bzl (hunk 6 above)
Comment (andrea-caforio) [reply rules/otbn.bzl:260]: I removed the `tpl` input parameter again. The command line is now the same as before.

Comment (etterli) [reply rules/otbn.bzl:260]: Ah now I understand how it works. It still generates a .s and .exp file for a testcase. All good then.

File: sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_test.hjson.tpl
Comment (sasdf) [inline sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_test.hjson.tpl:1]: I think this file can be removed.

File: sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+        m >>= 1
+    return y
+
+
+@testcase
+def gen_ntt_test(seed: Optional[int]):
+    # Generate random operands.
+    if seed is not None:
+        random.seed(seed)
```
Comment (sasdf) [inline sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py]: we can also move this rng setup to shared testcase

Comment (andrea-caforio) [reply sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py]: Done.

File: sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_test.hjson.tpl
Comment (andrea-caforio) [reply sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_test.hjson.tpl:1]: True.

File: rules/otbn.bzl [hunk 8]
```diff
@@ -217,6 +217,21 @@ def otbn_sim_test_suite(name, tests, **kwargs):
...
     )
 
+def otbn_sim_testgen(name, srcs, testgen, number, type = "expectation"):
+    """Bazel macro to automatically generate `number` of testcases.
+
+    An index i for (0 <= i < `number`) will be appended to the test name.
+    """
+
+    for i in range(number):
+        otbn_autogen_sim_test(
```
Comment (sasdf) [inline rules/otbn.bzl:229]: minor: it's useful for manual run if this macro a lso defines a manual test_suite with `name` , containing all the test it created.

Comment (andrea-caforio) [reply rules/otbn.bzl:229]: Done.

File: sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                y[j + m] = (y[j] - t) % Q
+                y[j] = (y[j] + t) % Q
+            start = m + (j + 1)
+
+        m >>= 1
+    return y
+
+
+@testcase
+def gen_ntt_test():
```
Comment (sasdf) [inline sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py]: Let's keep the interface taking seed (or args) for extensibility, although its unused in this test. 

e.g. useful for non-random predefined corner cases.

Comment (andrea-caforio) [reply sw/otbn/crypto/mldsa87/tests/mldsa87_ntt_testgen.py]: Done.
